MIDI Editor/talagan_OneSmallStep.lua

```diff
@@ -0,0 +1,395 @@
+    end
+
+    reaper.ImGui_PopStyleVar(ctx,2);
+
+    -- Separator
+    reaper.ImGui_NewLine(ctx);
+    ImGui_VerticalSpacer(ctx,5);
+
+    -- Note length line
+    reaper.ImGui_Text(ctx, "Note Length");
+    reaper.ImGui_SameLine(ctx);
+    reaper.ImGui_PushStyleVar(ctx, reaper.ImGui_StyleVar_FramePadding(),      0, 0);
+    reaper.ImGui_PushStyleVar(ctx, reaper.ImGui_StyleVar_ItemSpacing(),       2, 4);
+    reaper.ImGui_PushStyleVar(ctx, reaper.ImGui_StyleVar_ItemInnerSpacing(),  0, 0);
+
+    local nl = engine_lib.getNoteLen();
+    for s,k in ipairs({ "1", "1_2", "1_4", "1_8", "1_16", "1_32", "1_64" }) do
+      ButtonGroupImageButton('note_'.. k, nl==k,
+        function()
+          engine_lib.setNoteLen(k)
+        end
+      );
+    end
+
+    reaper.ImGui_SameLine(ctx);
+    reaper.ImGui_Dummy(ctx,10,0);
+    reaper.ImGui_SameLine(ctx);
+
+    local nmod = engine_lib.getNoteLenModifier();
+
+    ButtonGroupImageButton('note_dotted', nmod == engine_lib.NoteLenModifier.Dotted, function()
+        if nmod == engine_lib.NoteLenModifier.Dotted then
+          engine_lib.setNoteLenModifier(engine_lib.NoteLenModifier.Streight);
+        else
+          engine_lib.setNoteLenModifier(engine_lib.NoteLenModifier.Dotted);
+        end
+      end
+    );
+    ButtonGroupImageButton('note_triplet', nmod == engine_lib.NoteLenModifier.Triplet, function()
+        if nmod == engine_lib.NoteLenModifier.Triplet then
+          engine_lib.setNoteLenModifier(engine_lib.NoteLenModifier.Streight);
+        else
+          engine_lib.setNoteLenModifier(engine_lib.NoteLenModifier.Triplet);
+        end
+      end
+    );
+
+    reaper.ImGui_PopStyleVar(ctx,3);
+
+    if reaper.ImGui_IsWindowFocused(ctx) then
+      if not focustimer or reaper.ImGui_IsAnyMouseDown(ctx) then
+        -- create or reset the timer when there's activity in the window
+        focustimer = reaper.time_precise();
+      end
+
+      if reaper.time_precise() - focustimer > 0.5 then
+        local hwnd = reaper.GetMainHwnd();
+        reaper.JS_Window_SetFocus(hwnd)
+      end
+    else
+      focustimer = nil;
+    end
+
+    -- End
+    reaper.ImGui_End(ctx);
+
+  end
+
+  reaper.ImGui_PopStyleVar(ctx);
+
+
+  if open then
+    reaper.defer(loop)
+  else
+    stop();
+  end
+end
+
+
+function updateToolbarButtonState(v)
+  local _,_,sectionID,cmdID,_,_,_ = reaper.get_action_context();
+  reaper.SetToggleCommandState(sectionID,cmdID,v);
+  reaper.RefreshToolbar2(sectionID, cmdID);
+end
+
+function loop()
+  local engine_ret = engine_lib.atLoop();
+
+  if engine_ret == -42 then
+    reaper.ShowMessageBox("Could not install One Small Step's helper FX on the track.\n\nIf you just installed One Small Step, please try to restart REAPER to let it refresh its JFSX repository.", "Oops !", 0);
+    return;
+  end
+
+  ui_loop();
+end
+
+function stop()
+  updateToolbarButtonState(0);
+  reaper.ImGui_DestroyContext(ctx);
```

Comment:
This function no-op currently and will removed in ReaImGui 0.9, I highly recommend enabling the backward-compatibility module (imgui.lua) to shield the script against such breaking changes.

Reply:
The compatibility module is now enabled.